File: my_stuff/tcp_tsoding/listen.c
#include <stdio.h>
#include <stdlib.h>

#include <arpa/inet.h>
#include <sys/socket.h>
#include <unistd.h>

char buffer[4096];

int main() {
  struct sockaddr_in local;
  local.sin_family = AF_INET;
  inet_pton(AF_INET, "0.0.0.0", &local.sin_addr);
  local.sin_port = htons(6969);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  bind(sock, (struct sockaddr *)&local, sizeof(local));
  listen(sock, 128);

  int client = accept(sock, NULL, NULL);
  write(client, "220 Welcome\r\n", 13);

  int bytesRead = 0, res;
  for (;;) {
    res = read(client, buffer, sizeof(buffer));
    if (res < 0) {
      perror("reading");
      break;
    }
    if (!res)
      break;
    bytesRead += res;
  }

  printf("%d\n", bytesRead);
  close(sock);

  return 0;
}

File: my_stuff/tcp_tsoding/src/bin/listen.rs
use std::{
    io::{Read, Result, Write},
    net::TcpListener,
};

fn main() -> Result<()> {
    let addr = "0.0.0.0:6969";
    let listener = TcpListener::bind(addr)?;

    let (mut stream, _) = listener.accept()?;
    stream.write_all(b"220 Welcome\r\n")?;

    let mut buffer = [0; 4096];
    let mut total_bytes_read = 0;

    loop {
        let bytes_read = stream.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        total_bytes_read += bytes_read;
    }

    println!("{}", total_bytes_read);
    Ok(())
}
